### ISSUE TITLE
Incorrect "redundant cast" between total and partial dicts

### ISSUE THREAD

**Issue by alicederyn** [python/mypy#12200]
**Bug Report**

Casting a total dict to an equivalent non-total dict incorrectly warns that the cast is redundant.

**To Reproduce**

```python
class Permissive(TypedDict, total=False):
  id: str

class Restrictive(TypedDict):
  id: str

def convert(d: Restrictive) -> Permissive:
  return cast(Permissive, d)
```

**Expected Behavior**

No mypy errors/warnings

**Actual Behavior**

```
error: Redundant cast to "Permissive"
```

The mypy docs state "a total TypedDict is not valid when a partial one is expected", so this seems incorrect, and indeed if I remove the cast, I get:

```
Incompatible return value type (got "Restrictive", expected "Permissive")
```

**Your Environment**

- Mypy version used: Checked with all versions between 0.812 and 0.931 inclusive
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files):
```
python_version = "3.9"
namespace_packages = true
explicit_package_bases = true
strict = true
```
- Python version used: 3.9
- Operating system and version: docker, looks like RHEL 7.9, running within Ubuntu 10 on WSL 2

**Comment by alicederyn** [comment]
Apologies if this is a duplicate! I searched but the existing issues I found seemed like they were about other redundant cast bugs.

**Comment by A5rocks** [comment]
Note that it should be a redundant cast under #12142... I'm not sure if it will be accepted though.

**Comment by sodul** [comment]
We recently ran in a similar issue with boto stubs where strict output data from boto calls could not be re-used as input data for other boto calls where missing keys are allowed.

https://github.com/youtype/mypy_boto3_builder/discussions/208

The fix was apparently to remove the Output types and only use the types with `Total=False`, but this is a bit of a lie since the output types are supposed to contain all the keys.

Is there are better alternative than using `cast()` or always use `total=False` even if untrue?

**Comment by erictraut** [comment]
It appears this was fixed in mypy 0.981.

**Comment by hauntsaninja** [comment]
Fixed in #13311
